### PR TITLE
Clean up use of currlevel

### PR DESCRIPTION
**Pull request by AJenbo**
Hellfire set the wrong spawn position when entering crypt level 1:
image

This is fixed by this PR

### CODE REVIEW

File: Source/drlg_l1.cpp
Comment (AJenbo) [inline Source/drlg_l1.cpp:2253]: Previously this might actually have produced incorrect Crypt layouts for games that where converted from D1 to Hellfire (DevilutionX only). I see this as a self-bugfix despite being a possible break in comparability :shrug:

File: Source/drlg_l1.cpp
Comment (ephphatha) [inline Source/drlg_l1.cpp:2208]: When `PlaceMiniSet` is called with `tmax - tmin == 0` it calls the global rng exactly twice. This could use an early return like:
```suggestion
			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, true, -1) < 0) {
				// Skip trying to place the down stairs since we've already failed to generate the level, but we still need to advance the RNG state twice to match vanilla dungeon generation.
				AdvanceRndSeed();
				AdvanceRndSeed();
				return true;
			}
			if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
				return true;
```

**Review by ephphatha — COMMENTED** [review]
Would be good to get another set of eyes over the dungeon generation code. Breaking up those top level if statements into separate functions is good and looks like it hasn't changed anything from a read through, I just can't be confident I haven't missed anything 😅

File: Source/drlg_l1.cpp
Comment (AJenbo) [reply Source/drlg_l1.cpp:2196]: Yeah, I decided to wait with it til later not to risk breaking anything, and hopefully have written some unit tests first

File: Source/drlg_l1.cpp
Comment (AJenbo) [reply Source/drlg_l1.cpp:2325]: I had actually misunderstood this loop as continually adding to the level until there was no more room for set piece. That also makes refactoring it in to doing early returns much safer, thanks for figuring that out :+1:

File: Source/drlg_l1.cpp
Comment (AJenbo) [reply Source/drlg_l1.cpp]: There are also two entrances that are programmed incorrectly in the original Hellfire, one gets overwritten, the other case can never trigger, so there was never any real consequence.

**Comment by AJenbo** [comment]
@ephphatha thanks for pointing out the intention of the `GenerateLevel()` loop. I think `Place*MiniSets()` actually turned out to be sensible now :)

File: Source/drlg_l4.cpp
Comment (ephphatha) [inline Source/drlg_l4.cpp:1286]: each version of this function can be called something like `PlaceStairs`, bit more descriptive since they all deal with a specific set of minisets.

Comment (AJenbo) [reply Source/drlg_l4.cpp:1286]: Damn, thats exactly what I called it when i started on drlg_l4, but decided to rename it back to `PlaceMiniSets` :D

Comment (AJenbo) [reply Source/drlg_l4.cpp]: Fixed and rebased

**Comment by AJenbo** [comment]
Do you have a save for that?

File: Source/drlg_l1.cpp
Comment (AJenbo) [reply Source/drlg_l1.cpp:2169]: No worries I will just rewrite it again. Also nice catch :)

Sure would have been nice if they had put extra effort in to writing the level generation code well considering how important it is.

Comment (AJenbo) [reply Source/drlg_l1.cpp]: Alright, looks like only Cathedral/Crypt was affected. From them I removed the early return, except in the case that we are placing the stairs up in the none-Hellfire variant (this doesn't look as bad as it sounds I think :D ).

Diff from before: https://github.com/diasurgical/devilutionX/compare/7377ab32c8a7a430eede4ae24ec0fdb226fa77a0..8d486cc490910259838158851edb65dad47e4db4

**Comment by AJenbo** [comment]
Ugh, something is wrong about Catacombs as well:

image

**Comment by AJenbo** [comment]
Ok, fixed, but can someone explain to my why this:
```cpp
bool doneflag;
do {
	doneflag = PlaceStairs(entry);
} while (!doneflag);
```

is not the same as this:
```cpp
do {
} while (!PlaceStairs(entry));
```

File: Source/drlg_l1.cpp
Comment (AJenbo) [inline Source/drlg_l1.cpp:2173]: The original actually sets the wrong position for quest level entrances. Instead `GetReturnLvlPos()` will overwrite it with the position where the player was when triggering the level translation.

We could get rid of one of them. Any thoughts?

File: Source/drlg_l2.cpp
Comment (ephphatha) [inline Source/drlg_l2.cpp:2909]: You could even do this as `while (true) { if (PlaceStairs()) { break; } }` :D

would remove the need for `doneFlag`

Comment (qndel) [reply Source/drlg_l2.cpp:2909]: why not for(;;)
:D

**Review by ephphatha — COMMENTED** [review]
Ran through each level type and didn't spot any differences between master/this branch. The code is much nicer to read as well. Consider this approval, just wanted to give obligaron/qndel a chance to check before auto-merge kicks in.

File: Source/drlg_l2.cpp
Comment (AJenbo) [reply Source/drlg_l2.cpp]: @ephphatha done, I also flattened the one for caves.

File: Source/drlg_l1.cpp
Comment (AJenbo) [reply Source/drlg_l1.cpp:2173]: I have corrected it to give the correct position for PosionedWater, but I realize that it will also other adjustments for the other quests as most a positioned using .dun sections, except for laz portal.

**Review by obligaron — APPROVED** [review]
lgtm 🙂

**Comment by AJenbo** [comment]
Feels good to get to this point. I think we should be in good shape for handling custom Hellfire levels at this point :)

**Pull request merged**
